### ISSUE TITLE
Source Fetching

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#7158]
## Problem

Package management requires us to fetch the sources of packages and opam repositories. In particular, we need to understand the opam syntax for specifying package archives and pins.

## Solution

Write a library that will fetch package sources.

Implementation wise, fetching sources can be done in a few ways:

1. Fetch the packages with `$ opam source`
2. Use curl, git, and other tools to do the fetching.
3. For HTTP, vendor an HTTP/SSL stack in dune instead of using curl.

For v1, I suggest we just do the simplest thing (either 1. or 2.)

Eventually, we'll also need to be smart about source fetching and keep a cache of fetched sources like opam

**Comment by samoht** [comment]
Another option is to reuse `opam-lib` to do this, like it's done in [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/main/lib/pull.ml).

**Comment by rgrinberg** [comment]
> Another option is to reuse `opam-lib` to do this, like it's done in [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/main/lib/pull.ml?rgh-link-date=2023-02-22T17%3A32%3A22Z).

Yup, I mentioned this option on slack. Although it has a caveat:

> Feel free to reuse opam as a library for this as well. Just make sure that you’re using our process spawning and concurrency primitives

**Comment by Leonidas-from-XIV** [comment]
I've started writing the support for source fetching. The opam libraries have pretty decent support for this that we used in opam-monorepo before successfully.

The bad thing is that they return a `OpamProcess.Job.Op.job` which then runs on its own. The upside is that the types of that are exposed so I can implement the evaluation of a `job` to use `Fiber` instead. What's the preferred way to spawn processes in Dune? I've seen there is a vendored copy of `Spawn` but was wondering if there is some wrapper that lifts it into a `Fiber`?
